src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+/**
+ * (c) Copyright Reserved EVRYTHNG Limited 2021. All rights reserved.
+ * Use of this material is subject to license.
+ * Copying and unauthorised use of this material strictly prohibited.
+ */
+
+/* eslint-disable no-useless-escape */
+
+/**
+ * it protects some character with '%..'
+ * @param {string} input - the string to protect
+ * @return {string} the updated string
+ */
+export const webUriPercentEncoder = (input) => input.split('!').join('%21')
+  .split('(').join('%28')
+  .split(')')
+  .join('%29')
+  .split('*')
+  .join('%2A')
+  .split('+')
+  .join('%2B')
+  .split(',')
+  .join('%2C')
+  .split(':')
+  .join('%3A');
+
+/**
+ * @param {string} input - id that requires checkdigit
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @return {number} the checkDigit of the parameter
+ */
+const getCheckDigit = (input, throwError) => {
+  try {
+    const array = input.split('').reverse();
+    let total = 0;
+    let i = 1;
+    array.forEach((number) => {
+      const n = parseInt(number, 10);
+      if (i % 2 === 0) {
+        total += n;
+      } else {
+        total += n * 3;
+      }
+      i += 1;
+    });
+    return Math.ceil(total / 10) * 10 - total;
+  } catch (e) {
+    if (throwError) {
+      throw new Error(`There was a problem during the check digit operation of ${input}`);
+    }
+    return 0;
+  }
+};
+
+/**
+ * it adds the check sum at the end of the input and return a zero-padded result with a pre-defined
+ * length
+ * @param {string} raw - the string to convert
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @param length - the length of the output string
+ * @return {string} the updated string
+ */
+export const addCheckDigitAndZeroPad = (raw, throwError, length = 14) => {
+  let newRaw = raw;
+  newRaw += getCheckDigit(newRaw, throwError);
+  while (newRaw.length < length) {
+    newRaw = `0${newRaw}`;
+  }
+  return newRaw;
+};
+
+/**
+ * Converts any standard URI conveying a GS1 Key in Canonical GS1 DL URI.
+ * @param originalUri - any URI to be used in EPCIS events that convey a GS1 key, i.e. EPC URIs, EPC
+ * Class URIs, EPC ID Pattern URIs, or GS1 Digital Link URIs.
+ * @param {boolean} throwError - if set to true, it will throw an error if the event misses some
+ * fields for example. Otherwise, it won't throw an error and it will still return the generated id
+ * @return a corresponding, constrained version of a canonical GS1 Digital Link URI, i.e. with the
+ * lowest level of identification and without CPV/query string.
+ */
+export const dlNormalizer = (originalUri, throwError = true) => {
```

Comment:
This function is _gigantic_. Please consider refactoring it's many `if` cases to smaller, clearer separate functions.